designer/server/src/main/scala/pl/touk/nussknacker/ui/api/ScenarioLiveDataApiHttpService.scala

```diff
@@ -64,10 +69,55 @@ class ScenarioLiveDataApiHttpService(
                 Future.successful(Left(LiveDataNotSupported))
             }
           }
-        } yield LiveDataDto.from(liveData)
+        } yield LiveDataDto.from(liveData, computeCounts(scenarioWithDetails, liveData))
       }
   }
 
+  private def computeCounts(scenarioWithDetails: ScenarioWithDetails, liveData: LiveData)(
+      implicit loggedUser: LoggedUser
+  ): Map[String, NodeCount] = {
+    val uniqueNodeIds: Set[String] = liveData.nodeTransitions.keys.flatMap { t =>
+      t.destinationNodeId match {
+        case Some(dest) => List(t.sourceNodeId, dest)
+        case None       => List(t.sourceNodeId)
+      }
+    }.toSet
+
+    val incomingCounts =
+      liveData.nodeTransitions.toList
+        .flatMap { case (transition, data) => transition.destinationNodeId.map(_ -> data.totalCount) }
+        .groupBy(_._1)
+        .map { case (nodeId, counts) => nodeId -> counts.map(_._2).sum }
+
+    val outgoingCounts =
+      liveData.nodeTransitions.toList
+        .map { case (transition, data) => transition.sourceNodeId -> data.totalCount }
+        .groupBy(_._1)
+        .map { case (nodeId, counts) => nodeId -> counts.map(_._2).sum }
+
+    // We calculate counts based on transitions incoming to node, but if there are none (for sources) we use outgoing counts
+    val nodeCounts = uniqueNodeIds.map { nodeId =>
```

Comment:
`.map` on `Set` creates a new `Set` (computes all hashcodes), which is expensive, and we don't need to pay for this. To avoid this, you can use `.toList`. Also, please move `uniqueNodeIds` closer to this line to emphasize that this structure is needed for this purpose only.

Reply:
I changed it to use .distinct on list